get: handle non-DVC repositories

fabiosantoscode opened the pull request:
Allows us to `dvc get` from non-DVC source repositories.

Fixes #3089
Closes #3077

Will send PR for doc site tomorrow. Just need to change some parts which imply that the remote repository needs to be DVC-enabled.

----

* [x] ❗ Have you followed the guidelines in the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) list?

* [x] 📖 Check this box if this PR **does not** require [documentation](https://dvc.org/doc) updates, or if it does **and** you have created a separate PR in [dvc.org](https://github.com/iterative/dvc.org) with such updates (or at least opened an issue about it in that repo). Please link below to your PR (or issue) in the [dvc.org](https://github.com/iterative/dvc.org) repo.

docs: https://github.com/iterative/dvc.org/issues/911

* [x] ❌ Have you checked DeepSource, CodeClimate, and other sanity checks below? We consider their findings recommendatory and don't expect everything to be addressed. Please review them carefully and fix those that actually improve code or fix bugs.

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

fabiosantoscode commented:
DeepSource is complaining about some unused arguments in tests. I suppose I don't want to clean them up at this point.

efiop commented on dvc/repo/get.py:
```diff
@@ -43,7 +49,8 @@ def get(url, path, out=None, rev=None):
     tmp_dir = os.path.join(dpath, "." + str(shortuuid.uuid()))
     try:
-        with external_repo(cache_dir=tmp_dir, url=url, rev=rev) as repo:
+        cached_clone(url, rev=rev, clone_path=tmp_dir)
```
You don't need to pass tmp_dir to cached_clone. We do that in `external_repo` because we want our cache to be located right beside the location that we want to checkout it to to speed it up and have an ability to use reflinks and hardlinks, instead of copying from another filesystem. Btw, you've lost that optimization in your change.

fabiosantoscode replied:
I thought this was enough, since the Repo is rooted at `tmp_dir` I suppose it's being cloned right there, and then copied from the same FS? What am I missing?

efiop replied:
That is a good point! I've missed that! Indeed, when doing so you no longer need to change cache dir. The reason we do that is that there is a chance that the fs that we are currently on doesn't have proper support for some stuff like locking (e.g. https://www.sqlite.org/howtocorrupt.html#_filesystems_with_broken_or_missing_lock_implementations), which might make `Repo()` freeze. So it is best to keep it the way it was before.

efiop commented on dvc/repo/get.py:
```diff
@@ -53,26 +60,24 @@ def get(url, path, out=None, rev=None):
+            output = repo.find_out_by_relpath(path)
+            if not output.use_cache:
+                # Catch this below and go for a plain old fs_copy
+                raise _DoPlainCopy
```
This is a very strange flow control, I have never ever seen one like that, but maybe I haven't seen enough python projects, so please correct me if I'm wrong. Regardless, I pretty sure you could simply do something like:
```
try:                                                                    
    repo = Repo(tmp_dir)                                                
    output = repo.find_out_by_relpath(path)                             
    if output.use_cache:                                                
        ... 
        _get_cached(repo, output, out)                                  
        return                                                          
except (NotDvcRepoError, OutputNotFoundError):                          
    pass                                                                
                                                                        
if os.path.isabs(path):                                                 
    raise FileNotFoundError                                             
                                                                        
fs_copy(os.path.join(tmp_dir, path), out)                               
```
and avoid that.

fabiosantoscode replied:
Yes, that is way nicer!

efiop commented on dvc/repo/get.py:
```diff
@@ -53,26 +60,24 @@ def get(url, path, out=None, rev=None):
             #
             # Also, we can't use theoretical "move" link type here, because
             # the same cache file might be used a few times in a directory.
```
This comment should stay over `repo.cache.local.cache_types = ["reflink", "hardlink", "copy"]` line, otherwise it looses some of its meaning.

fabiosantoscode replied:
Oops

efiop commented:
> DeepSource is complaining about some unused arguments in tests. I suppose I shouldn't clean them up at this point.

@fabiosantoscode Yep, take it as a heads-up from DS. In this case there are no actions you should take. :slightly_smiling_face:

fabiosantoscode commented:
@efiop I believe I've addressed the comments.

WRT the `clone_dir` argument and the lost optimization, I removed the argument, and instead of changing the clone target, I changed the Repo's cache to `tmp_dir` which is next to the place the `get` will drop the file at.

efiop commented on dvc/repo/get.py:
```diff
@@ -55,24 +64,24 @@ def get(url, path, out=None, rev=None):
-            if output and output.use_cache:
+            output = repo.find_out_by_relpath(path)
+            if output.use_cache:
+                # Catch this below and go for a plain old fs_copy
```
```suggestion
```

efiop commented on dvc/repo/get.py:
```diff
@@ -28,8 +29,15 @@ def __init__(self):
+class _DoPlainCopy(DvcException):
+    pass
+
+
```
No longer need that.

fabiosantoscode commented:
Cool @efiop, I've reverted the code to just clone into `tmp_dir` and addressed the remaining comments :+1:

efiop commented:
> WRT the clone_dir argument and the lost optimization, I removed the argument, and instead of changing the clone target, I changed the Repo's cache to tmp_dir which is next to the place the get will drop the file at.

Sorry, not sure I follow. I see that you still pass tmp_path to cached_clone , instead of keeping it as it was before and only changing the cache dir.

fabiosantoscode commented:
@efiop I wrote that before you pointed out that it was enough to just clone the repo to `tmp_dir`, do disregard it.

jorgeorpinel commented on dvc/command/get.py:
```diff
@@ -31,7 +31,7 @@ def run(self):
 
 def add_parser(subparsers, parent_parser):
-    GET_HELP = "Download/copy files or directories from DVC repository."
+    GET_HELP = "Download/copy files or directories from Git repository."
```
Maybe let's use
```suggestion
    GET_HELP = "Download a file or directory from any DVC project or Git repository."
```
to match https://dvc.org/doc/command-reference/get

jorgeorpinel replied:
Maybe let's use
```suggestion
    GET_HELP = "Download a file or directory from any DVC project or Git repository"
```
to match https://dvc.org/doc/command-reference/get

jorgeorpinel commented on dvc/command/get.py:
```diff
@@ -40,10 +40,10 @@ def add_parser(subparsers, parent_parser):
     )
     get_parser.add_argument(
-        "url", help="URL of Git repository with DVC project to download from."
+        "url", help="URL of Git repository to download from."
```
```suggestion
        "url", help="Location of DVC project or Git repository to download from"
```
p.s. it can simply be a local file system path, even to a non-Git DVC project dir, I think.

fabiosantoscode replied:
The fact this is called "url" might be confusing because of that.

jorgeorpinel replied:
Yeah and also `path` since URL can be a path too... They're edge cases and explained in the docs so it should be OK.

jorgeorpinel commented on dvc/command/get.py:
```diff
@@ -40,10 +40,10 @@ def add_parser(subparsers, parent_parser):
     )
     get_parser.add_argument(
-        "path", help="Path to a file or directory within a DVC repository."
+        "path", help="Path to a file or directory within the repository."
```
```suggestion
        "path", help="Path to a file or directory within the project or repository"
```

jorgeorpinel commented on dvc/command/get.py:
```diff
@@ -52,6 +52,6 @@ def add_parser(subparsers, parent_parser):
     )
     get_parser.add_argument(
-        "--rev", nargs="?", help="DVC repository git revision."
+        "--rev", nargs="?", help="Repository git revision."
```
```suggestion
        "--rev", nargs="?", help="Git revision (e.g. branch, tag, SHA)"
```

fabiosantoscode commented:
@jorgeorpinel thanks for your suggestions, it reads a lot clearer now.

As per fixing #3077, sure, let's do it :)

fabiosantoscode commented:
@jorgeorpinel the docs for import are now updated. Check them out and see if they make sense to you.

@efiop your comments have all been addressed, I believe.

And @restyled-io has closed its PR, since the code style is on par. I ran `black dvc` locally just to confirm.

fabiosantoscode commented:
I'll update the docs when this is confirmed to be the final help text.

efiop commented on dvc/repo/get.py:
```diff
@@ -43,7 +44,10 @@ def get(url, path, out=None, rev=None):
     tmp_dir = os.path.join(dpath, "." + str(shortuuid.uuid()))
     try:
-        with external_repo(cache_dir=tmp_dir, url=url, rev=rev) as repo:
+        cached_clone(url, rev=rev, clone_path=tmp_dir)
```
Sorry, I might've not worded it clearly enough, but I meant that we should keep it as it was before this change, meaning that we shouldn't pass tmp_dir to cached_clone and let it clone wherever it needs to. And we need to setup external cache dir pointing to tmp_dir, the same way we do that in external_repo. Also, take a look at `external_repo`, it is doing things like setting a default remote, which you've lost here.

efiop replied:
You might be able to flip the logic over and do something like:
```
try:
    with external_repo():
        ...
        if out.use_cache:
            _get_cached()
           return
except NotDvcRepoError:
     pass
path = cached_clone(...)
# copy from git
```

fabiosantoscode replied:
Done! I didn't fold it like that because I would have to do the cached_clone conditionally, and instead duplicated the file copy call.

jorgeorpinel commented on dvc/command/get.py:
```diff
@@ -40,10 +42,12 @@ def add_parser(subparsers, parent_parser):
     get_parser.add_argument(
-        "url", help="URL of Git repository with DVC project to download from."
+        "url",
+        help="Location of DVC project or Git repository to download from.",
```
Not a complete sentence so no period
```suggestion
        help="Location of DVC project or Git repository to download from",
```

jorgeorpinel commented on dvc/command/imp.py:
```diff
@@ -41,13 +42,12 @@ def add_parser(subparsers, parent_parser):
     import_parser.add_argument(
-        "url", help="URL of Git repository with DVC project to download from."
+        "url",
+        help="Location of Git repository with DVC project to download from.",
```
```suggestion
        help="Location of DVC project or Git repository to download from",
```
same as `dvc get`

jorgeorpinel commented on dvc/command/imp.py:
```diff
@@ -41,13 +42,12 @@ def add_parser(subparsers, parent_parser):
+        "url",
+        help="Location of Git repository with DVC project to download from.",
     )
+    import_parser.add_argument("path", help="Path to data within DVC project.")
```
```suggestion
    import_parser.add_argument("path", help="Path to a file or directory within the project or repository")
```
same as `dvc get`

jorgeorpinel commented on dvc/command/get.py:
```diff
@@ -52,6 +56,6 @@ def add_parser(subparsers, parent_parser):
         help="Destination path to copy/download files to.",
```
```suggestion
        help="Destination path to download files to",
```
since we're no longer using the term "copy" in these texts

jorgeorpinel replied:
Not sure why GH thinks my suggestion removes a space character. It doesn't. That's weird. Oh well...

jorgeorpinel commented on dvc/command/imp.py:
```diff
@@ -41,13 +42,12 @@ def add_parser(subparsers, parent_parser):
-    )
-    import_parser.add_argument(
-        "-o", "--out", nargs="?", help="Destination path to put data to."
+        "-o", "--out", nargs="?", help="Destination path to put data in."
```
```suggestion
        "-o", "--out", nargs="?", help="Destination path to download files to"
```
same as `dvc get`

jorgeorpinel commented on dvc/command/imp.py:
```diff
@@ -41,13 +42,12 @@ def add_parser(subparsers, parent_parser):
+        "-o", "--out", nargs="?", help="Destination path to put data in."
     )
     import_parser.add_argument(
         "--rev", nargs="?", help="DVC repository git revision."
```
```suggestion
        "--rev", nargs="?", help="Git revision (e.g. branch, tag, SHA)"
```
same as `dvc get`

fabiosantoscode commented:
My bad @jorgeorpinel I should've made the help texts for get and import uniform myself. Thanks for doing that for me in the end.

@efiop now I'm using `external_repo` like before, to leverage its DVC-specific logic, and a simple cached_clone followed by a file copy in case it's not a DVC repo.

I'm wondering if a simple file move would be the best for the simple git repository case. After all, the repository is destroyed just after and moves within a single filesystem are very fast.

shcheklein commented on dvc/repo/get.py:
```diff
@@ -42,39 +41,50 @@ def get(url, path, out=None, rev=None):
+                    return
+
                 # Either an uncached out with absolute path or a user error
+
```
minor: remove this extra line?

fabiosantoscode replied:
:+1:

efiop commented on dvc/repo/get.py:
```diff
@@ -42,39 +41,49 @@ def get(url, path, out=None, rev=None):
+                    return
+
                 # Either an uncached out with absolute path or a user error
                 if os.path.isabs(path):
```
Absolute paths are not allowed even for pure-git repos, hence why I've suggested that example that has the same code handling both non-cached and git files. No need to worry much about second cached_clone being redundand, as we cache those anyways, so the overhead is not big.

fabiosantoscode replied:
Got it. I tried to fold it like you demonstrated before, but it looks like the files are missing if I try to do the copy outside of the context manager.

So I extracted the part that checks if the path is absolute into a function and called it twice.

efiop commented on dvc/repo/get.py:
```diff
@@ -42,39 +41,49 @@ def get(url, path, out=None, rev=None):
     finally:
         remove(tmp_dir)
+        if raw_git_dir:
+            remove(raw_git_dir)
```
I don't think there is any need in removing it, it is a cached copy after all 🙂 And if you check external_repo, it doesn't delete those either. Or am I missing something?

fabiosantoscode replied:
That's good news. I figured there was some cleaning up happening but haven't seen how it works yet.

efiop commented on dvc/repo/get.py:
```diff
@@ -43,36 +47,43 @@ def get(url, path, out=None, rev=None):
 
                 fs_copy(os.path.join(repo.root_dir, path), out)
+                return
+
```
Why not just pass it through? You are doing the same thing down below anyway.

fabiosantoscode replied:
First time I tried it I got file not found errors and figured the context manager was removing the repository. I was wrong. I got it working now.

jorgeorpinel commented on dvc/command/get.py:
```diff
@@ -40,18 +42,17 @@ def add_parser(subparsers, parent_parser):
     get_parser.add_argument(
-        "url", help="URL of Git repository with DVC project to download from."
+        "url",
+        help="Location of DVC project or Git repository to download from",
```
Argh. I guess the term "Location" here is too confusing after all... I'll address this separately.

Suor commented on dvc/repo/get.py:
```diff
@@ -42,37 +41,47 @@ def get(url, path, out=None, rev=None):
     except (OutputNotFoundError, FileNotFoundError):
         raise PathMissingError(path, url)
-    except NotDvcRepoError:
-        raise UrlNotDvcRepoError(url)
```
I think we should revert this at least for api code.

efiop replied:
@Suor Not sure I follow, `get` is not even part of the api.

Suor replied:
This is making better error message for code using external_repo, but that exception is caught in get and import code, so these two lines won't fire for them, but it can still fire for any other external_repo user like api.

efiop replied:
@Suor Yes, but we are looking at `get` here, not `external_repo`, which wasn't wrapped anyways. @skshetry has noticed that too today and will be adding a wrapper for `api.get_url` soon too.

Suor replied:
Disregarding whether we have or don't have a wrapper for api, which has functions beside `get_url`, `external_repo()` should raise errors with appropriate messages. Otherwise we force very tricky interface on all `external_repo()` users they need to catch some exceptions, and if they forget too than users will get wrong error messages. 

Removing this lines make `external_repo()` abstraction broken. We should restore it and we won't need any wrapper in api code and make using `external_repo()` easier.

efiop replied:
@Suor Once again, you are talking about a general problem under a PR that added support for specific feature that doesn't affect the general problem. There are no action points in this PR, let's stop flooding it with off-topic stuff please.

fabiosantoscode commented:
@Suor maybe I don't understand what you've raised here but since get now supports non DVC repositories it will never throw an exception to complain about the target repository not being DVC.

Suor commented:
@fabiosantoscode get will never throw, but `external_repo()` doesn't know and should not expect all code using it to always handle certain exceptions. Anyway, this won't be an issue after #3190.
